File: haskell/isogram/src/Isogram.hs
module Isogram (isIsogram) where

import           Data.Char          (isAlpha, toLower)
import qualified Data.Set   as S


isIsogram :: String -> Bool
isIsogram str =
    length ls == S.size (S.fromList ls)
    where ls = toLower <$> filter isAlpha str

File: c/isogram/src/isogram.c
#include <ctype.h>
#include <stdlib.h>
#include "isogram.h"

bool is_isogram(const char phrase[])
{
    if (phrase == NULL)
    {
        return false;
    }

    // lettersFound[i] = 1 if the ith letter of the alphabet
    // has already been found in phrase, 0 otherwise.
    int lettersFound[26] = {0};
    const char *c = phrase;
    while (*c != '\0')
    {
        if (isalpha(*c))
        {
            int i = tolower(*c) - 'a';
            if (lettersFound[i] != 0)
            {
                return false;
            }
            lettersFound[i] = 1;
        }
        c++;   
    }

    return true;
}